prov/shm/src/smr_atomic.c

```diff
@@ -308,12 +308,14 @@ static ssize_t smr_atomic_writemsg(struct fid_ep *ep_fid,
 
 	ep = container_of(ep_fid, struct smr_ep, util_ep.ep_fid.fid);
 
+	if (!(ep->util_ep.cq_bind_flags & OFI_TX_CQ_SELECTIVE))
+		flags |= FI_COMPLETION;
```

Comment:
I don't like that this introduces a branch that gets copied multiple times, particularly, since we know before the function is even called what it should do.  I'm trying to find a decent alternative.

Reply:
I know, me too. The other option I can think of is to include 2 new fields in the util_ep
default_rx_completion and default_tx_completion that gets set to FI_COMPLETION if not bound with SELECTIVE_COMPLETION and gets set to 0 if it is. Instead of checking the flag on every send, we just OR it in for every operation. Does that make sense? Adds a bit more memory but gets rid of the branching.

Reply:
I thought about that as well.  I even came up with a way to use the u8 bit field you defined and apply a mask + bit shift that gets ORed in instead.  But that results in a complex calculation that I'm not convinced is better than the branch.

Another option would be to have a separate set of msg functions:
```
blah_msg_comp(...)
{
    return blah_msg(..., flags | FI_COMPLETION);
}
```

I think the psm2 provider might do something like this.

Reply:
It looks like psm2 has 2 ep fields 'send_selective_completion' and 'recv_selective_completion' and then on a send it checks the ep fields:
```
if ((flags & PSMX2_NO_COMPLETION) ||
     (ep_priv->send_selective_completion && !(flags & FI_COMPLETION)))
          no_completion = 1;
```